### DISCUSSION TITLE
Shorthand version of EventArgs primary constructor

### DISCUSSION THREAD

**Discussion by Booser** [dotnet/csharplang#8462] [Language Ideas]
When creating new EventArgs overloads I must write a code block like this.
`    public class AlertEventArgs(string message, string notificationid) : EventArgs
    {
        public string Message { get; } = message;
        public string NotificationId { get; } = notificationid;
    }
`
When I try to use the style like record types use it is nice and short but there where no generated properties in the lowered code.

`    public class AlertEventArgs(string Message, string Notificationid) : EventArgs;
`

Wouldn't it be nice if the second version creates the same experience much more compact? Maybe an extra Attribute to make it happen. Nice and clean.

https://sharplab.io/#gist:04c00f32a6c37fca899afb724076356d

**Comment by FaustVX** [comment]
What you're asking has nothing to do with `EventArgs`, it's just the primary constructor syntax in a class and `EventArgs` is nothing special here.

What you can do is use my [source generator package](https://www.nuget.org/packages/FaustVX.PrimaryParameter.SG) (or write your own) and just make the type `partial` with `[Property]` attribute on the primary parameters :
```C#
public partial class AlertEventArgs([Property]string message, [Property]string notificationid) : EventArgs;
```
Here is what will be generated at the end for you automatically :
```C#
public partial class AlertEventArgs : EventArgs
{
  public string Message { get; }
  public string Notificationid { get; }
  AlertEventArgs(string message, string notificationid)
  {
    Message = message;
    Notificaionid = notificationid;
  }
}
```

**Reply by Booser** [reply]
Thanks for the suggestion i'm testing it, but it would be nicer when it yould be included somehow.

**Reply by FaustVX** [reply]
That was in intentional decision from the Language Design Team to not generate members on types other than `record`

**Comment by huoyaoyuan** [comment]
It's basically inheriting record-like type from non-record class.

**Reply by colejohnson66** [reply]
I've always wondered why this restriction exists? A record is just a fancy class anyways.

**Reply by HaloFour** [reply]
> I've always wondered why this restriction exists? A record is just a fancy class anyways.

Mostly because it breaks (or would make it very difficult for the compiler to enforce) two of the primary features of records, namely value equality and "withers".  A non-record parent wouldn't participate in the `EqualityContract` and it would be a challenge to ensure that the parent class "does the right thing", otherwise the child record would end up with reference equality or something in between.  And with "withers", if that isn't supported by copy constructors all the way through the record type hierarchy, you end up with decapitation scenarios where some fields end up copied and others don't.

**Reply by colejohnson66** [reply]
Ah. Makes sense; I forgot about those features. So the best/only solution here would be a source generator then that generates properties from the primary constructor (such as @FaustVX's).

**Comment by CyrusNajmabadi** [comment]
A source-generator could add these members for you.  You'd just need to make the type partial.
